zallet/src/commands/migrate_zcash_conf.rs

```diff
@@ -0,0 +1,558 @@
+                datadir.join(&self.conf)
+            } else {
+                default_data_dir()
+                    .ok_or(ErrorKind::Generic)?
+                    .join(&self.conf)
+            }
+        } else {
+            self.conf.to_path_buf()
+        };
+
+        let f = BufReader::new(
+            File::open(&conf)
+                .await
+                .map_err(|e| ErrorKind::Generic.context(e))?,
+        );
+        let mut lines = f.lines();
+
+        let actions = build_actions();
+        let mut config = ZalletConfig::default();
+        let mut warnings = vec![];
+
+        while let Some(line) = lines
+            .next_line()
+            .await
+            .map_err(|e| ErrorKind::Generic.context(e))?
+        {
+            // Parse the Boost.Program_options file format.
+            if line.is_empty() || line.starts_with('#') {
+                continue;
+            }
+            let (option, rest) = match line.split_once('=') {
+                Some(res) => Ok(res),
+                None => Err(ErrorKind::Generic.context(fl!(
+                    "err-migrate-invalid-line",
+                    line = line,
+                    conf = conf.display().to_string(),
+                ))),
+            }?;
+            let value = rest
+                .split_once('#')
+                .map(|(value, _)| value.trim_end())
+                .unwrap_or(rest.trim_end());
+
+            match actions.get(option) {
+                Some(Action::MapTo(f)) => f(&mut config, value)?,
+                Some(Action::Ignore) => (),
+                Some(Action::Warn(f)) => {
+                    if let Some(warning) = f(value) {
+                        warnings.push(warning);
+                    }
+                }
+                None => {
+                    return Err(ErrorKind::Generic
+                        .context(fl!("err-migrate-unknown-zcashd-option", option = option,))
+                        .into())
+                }
+            }
+        }
+
+        // Inform the user of any warnings.
+        if !warnings.is_empty() {
+            println!("{}", fl!("migrate-warnings"));
+            for warning in warnings {
+                println!();
+                println!("{warning}")
+            }
+
+            // Warnings must be allowed by the user.
+            if !self.allow_warnings {
+                return Err(ErrorKind::Generic
+                    .context(fl!("err-migrate-allow-warnings"))
+                    .into());
+            }
+        }
+
+        if !self.this_is_alpha_code_and_you_will_need_to_redo_the_migration_later {
+            return Err(ErrorKind::Generic.context(fl!("migrate-alpha-code")).into());
+        }
+
+        // Serialize the config.
+        let config_toml =
+            toml::Value::try_from(config).map_err(|e| ErrorKind::Generic.context(e))?;
+        let output =
+            toml::to_string_pretty(&config_toml).map_err(|e| ErrorKind::Generic.context(e))?;
```

Comment:
Due to how `serde + toml` serialization works, any config option that is `Option<T>` is omitted when `None`, but any config option that is just `T` is not omitted when set to its default value. This is how `zebrad` works as well, but I'm not very happy with it. I'm half inclined to make _every_ config option `Option<T>` so that it can correctly represent "user-set / not-user-set" in memory, and then have getters that map `None` to the intended default value.

Reply:
I guess that seems reasonable. In some respects though I kind of like the idea that the defaults are explicitly set in the config file, because then the user doesn't have to go hunting in various places to find out what options are available and what their defaults are; they just get to see all the options up front when looking at the config file.

I know that I value when default config files are "complete" and self-documenting, such that I don't have to go looking for separate documentation.

Reply:
I think what I'd prefer is to have a separate command (similar to `zebrad generate`) that generates a config file containing _all_ of the defaults (unlike `zebrad` which doesn't fill in defaults for some of its values like `Option<T>`s). Then users can run it to get a readable form of whatever the current defaults are, and adapt / use them at will. (The downside of offering this command is that most users might end up configuring the defaults explicitly, preventing us from ever updating them. So maybe the command should actually generate an empty config file with every default commented out.)

For this command, I prefer that it only generate a file containing settings that were explicitly configured by the user in their `zcash.conf`.

Reply:
That makes sense. Having the defaults commented out is a pretty standard way of approaching the problem, and in the associated comments we can also describe *why* those particular defaults were chosen, and what the valid ranges etc. are.